src/main/java/net/dv8tion/jda/bot/entities/impl/DefaultShardManagerImpl.java

```diff
@@ -0,0 +1,336 @@
+        this.backoff = backoff;
+
+        if (shardsTotal != -1)
+        {
+            if (shardIds == null)
+            {
+                this.shards = new ShardCacheViewImpl(shardsTotal);
+                for (int i = 0; i < this.shardsTotal; i++)
+                    this.queue.offer(i);
+            }
+            else
+            {
+                this.shards = new ShardCacheViewImpl(shardIds.size());
+                shardIds.stream().distinct().sorted().forEach(this.queue::offer);
+            }
+        }
+    }
+
+    @Override
+    public void addEventListener(final Object... listeners)
+    {
+        Collections.addAll(this.listeners, listeners);
+        ShardManager.super.addEventListener(listeners);
+    }
+
+    @Override
+    public int getAmountOfQueuedShards()
+    {
+        return this.queue.size();
+    }
+
+    @Override
+    public ShardCacheView getShardCache()
+    {
+        return this.shards;
+    }
+
+    public void login() throws LoginException, IllegalArgumentException
+    {
+        // building the first one in the currrent thread ensures that LoginException and IllegalArgumentException can be thrown on login
+        JDAImpl jda = null;
+        try
+        {
+            if (this.queue.isEmpty())
+            {
+                jda = this.buildInstance(0);
+
+                this.shards = new ShardCacheViewImpl(this.shardsTotal);
+                for (int i = 0; i < this.shardsTotal; i++)
+                    this.queue.offer(i);
+
+                this.shards.getMap().put(0, jda);
+            }
+            else
+            {
+                final int shardId = this.queue.peek();
+
+                jda = this.buildInstance(shardId);
+                this.shards.getMap().put(shardId, jda);
+                this.queue.remove(shardId);
+            }
+        }
+        catch (final RateLimitedException e)
+        {
+            // do not remove 'shardId' from the queue and try the first one again after 5 seconds in the async thread
+        }
+        catch (final Exception e)
+        {
+            if (jda != null)
+                jda.shutdown();
+            throw e;
+        }
+
+        this.worker = this.executor.scheduleWithFixedDelay(() ->
+        {
+            if (this.queue.isEmpty())
+                return;
+
+            JDAImpl api = null;
+            try
+            {
+                int shardId = -1;
+                do
+                {
+                    final int i = this.queue.peek();
+                    if (this.shards.getMap().containsKey(i))
+                        this.queue.poll();
+                    else
+                        shardId = i;
+                }
+                while (shardId == -1 && !this.queue.isEmpty());
+
+                if (shardId == -1)
+                {
+                    this.queue.poll();
+                    return;
+                }
```

Comment:
Why do you poll here?

Reply:
This is a good question. If the code gets here the queue is empty anyway. Fixed.